fuzz/src/lib.rs

```diff
@@ -175,33 +201,140 @@ pub trait DbSimulator {
 					old_layers.push(layers.clone());
 					db = {
 						drop(db);
-						retry_operation(|| parity_db::Db::open(&options))
+						retry_operation(|| DbWithIter::open(&options))
 					};
-					Self::reset_model_from_database(&db, &mut layers, &old_layers);
+					Self::reset_model_from_database(&db.db, &mut layers, &old_layers);
 				},
+				Action::IterPrev =>
+					if let Some(iter) = &mut db.iter {
+						let mut old_key = if let Some(old_key) = db.iter_current_key.take() {
+							old_key
+						} else {
+							retry_operation(|| iter.seek_to_last());
+							IterPosition::End
+						};
+						let new_key_value =
+							iter.prev().unwrap_or_else(|e| {
+								log::debug!("Database error: {}, restarting iter.prev without I/O limitations.", e);
+
+								// We ignore the error and reset the iterator
+								parity_db::set_number_of_allowed_io_operations(usize::MAX);
+								iter.seek_to_last().unwrap();
+								old_key = IterPosition::End;
+								iter.prev().unwrap()
+							});
+						let mut content = Self::layer_required_content(&model_values(&layers));
+						content.sort();
+						match old_key {
+							IterPosition::Start => {
+								log::info!(
+									"Prev lookup on iterator in start state, expecting None"
+								);
+								assert_eq!(new_key_value, None);
+								db.iter_current_key = Some(IterPosition::Start);
+							},
+							IterPosition::Value(old_key) => {
+								let expected = Self::valid_previous_values(&[old_key], &layers);
+								log::info!(
+									"Prev lookup on iterator on old key {}, expecting one of {:?}",
+									old_key,
+									expected
+								);
+								assert!(expected.contains(&new_key_value), "Prev lookup on iterator on old key {}, expecting one of {:?}, found {:?}",
+										old_key,
+										expected, new_key_value);
+								db.iter_current_key =
+									Some(new_key_value.map_or(IterPosition::Start, |(k, _)| {
+										IterPosition::Value(k[0])
+									}));
+							},
+							IterPosition::End => {
+								let expected =
+									Self::valid_previous_values(&[u8::MAX, u8::MAX], &layers);
```

Comment:
Could be IterPosition::End (current code works because key space is a single u8, but  End would be better), by changing valid_previous_values to use IterPosition as parameter.
Same thing for valid_next_value call assuming the key are not 0 length
Edit: actually does not really matter

Reply:
Done. It allowed me to factorise quite well the code. Thank you for pushing me to do it!